pkg/ingest/conn.go

```diff
@@ -172,3 +175,39 @@ func (m *connectionManager) isEmpty() bool {
 	defer m.mtx.RUnlock()
 	return len(m.active) <= 0
 }
+
+// streamClock is a logical clock that is initialized to start at a random offset
+// that identifies a stream.
+// It always holds the next clock tick and is incremented when Read is called.
+// It's guaranteed that the 4 high bytes are 0 initially. If they overflow, the random
+// component is incremented by 1, effectively creating a new stream identifier.
+// It is not safe to be used concurrently.
+type streamClock [10]byte
+
+func newSteamClock() (*streamClock, error) {
+	var sc streamClock
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	if _, err := r.Read(sc[:6]); err != nil {
+		return nil, err
+	}
+	// Set a bit of the random part to 0 so we can increment the random part
+	// when the clock overflows.
+	// We pick the first bit of the second byte as it simplifies the Read implementation.
+	sc[2] &= 0x7f
+	return &sc, nil
+}
+
+// Read populates b with the next clock tick and increments the clock.
+// b must be of length 10.
+func (sc *streamClock) Read(b []byte) (int, error) {
+	if len(b) != 10 {
```

Comment:
Is it necessary for b to be exactly of len(10) or would it be sufficient for it to be at least that length?

Reply:
I don't think so.

We are doing this here to verify an assumption of the `ulid` packages implementation, i.e. it reads all randomness in one go and not `Read(3), Read(4), Read(3)`. If we allowed `Read(25)`, we'd probably leave the 2nd half at zero but had no way to know from which bytes the `ulid` package actually took it's entropy (could skip every 2nd byte for example).

It's certainly slightly sketchy overall to add an assumption the interface does not guarantee. But the strict check should add a reasonable amount of safety.

Reply:
This SGTM for the record.